cql3/values.hh

```diff
@@ -54,30 +55,39 @@ struct raw_value_view {
     // - pointers are cheap to copy
     // - it makes the view keep its semantics - it's safe to copy a view multiple times
     //   and all copies still refer to the same underlying data.
-    lw_shared_ptr<bytes> _temporary_storage = nullptr;
+    lw_shared_ptr<managed_bytes> _temporary_storage = nullptr;
 
-    raw_value_view(null_value&& data)
+    raw_value_view(null_value data)
         : _data{std::move(data)}
     {}
-    raw_value_view(unset_value&& data)
+    raw_value_view(unset_value data)
         : _data{std::move(data)}
     {}
     raw_value_view(fragmented_temporary_buffer::view data)
         : _data{data}
     {}
+    raw_value_view(managed_bytes_view data)
+        : _data{data}
```

Comment:
A possible optimization is to add bytes_view to the variant, and here (and in the temporary constructor) select the bytes_view branch if there are <= 1 fragments. This will bloat the generated code more (we'll have 9 versions of the equal() function), but create shorter paths for the very common case of a value fitting in a single fragment.

Let's defer it so it can come with some measurement that proves it helps.

Reply:
What do you mean by "shorter paths" in this case?

If anything, I could see the benefit of increased branch predictability (because with_value would almost always choose the same branch), and decreased icache usage (because the code paths for fragmented_temporary_view::view and managed_bytes_view would be almost completely cold).
But:
1. Branch predictability and icache impact would probably need specialized microbenchmarks to measure (because our existing ones are predictable in the first place).
2. The entire thing sounds like a "100 CPU cycles per query" type of deal.

Probably not worth it.

Reply:
Shorter paths means that we make a decision once in a higher level, then work with smaller and simpler data structures below.

There are three classes of workloads:
1. One which has mostly small values. Here, each cycle is important, because all of the work is spent on making decisions and the per-byte work is negligible.
2. One which has mostly large values. Here, there's no point in counting cycles since a lot of the work is moving large amounts of data around (we also expect higher miss rates from the row cache).
3. For completeness, a mixed workload.

My goal is to optimize for the small-value case since it's the one that's going to be CPU bound. The large-value and mixed-value workloads are going to be dominated by data movement (across the network and within the core). So branch prediction and instruction counts matter mostly for the small-value workload. So for the workload that benefits, it will predict the single-fragment case well.

It's not 25 cycles per query, but 25 cycles each time we touch a value, which can be dozens of times per query. It is still small compared to overall query cost (which can be 60,000+ cycles), but not negligible.